ddtrace/monkey.py

```diff
@@ -56,127 +46,96 @@
-# DEV: This ensures we do not patch a module until it is needed
-# DEV: <contrib name> => <list of module names that trigger a patch>
-_PATCH_ON_IMPORT = {
-    'celery': ('celery', ),
-    'flask': ('flask, '),
-    'gevent': ('gevent', ),
-    'requests': ('requests', ),
-}
-
 
 class PatchException(Exception):
-    """Wraps regular `Exception` class when patching modules"""
     pass
 
 
-def _on_import_factory(module, raise_errors=True):
-    """Factory to create an import hook for the provided module name"""
-    def on_import(hook):
-        # Import and patch module
-        path = 'ddtrace.contrib.%s' % module
-        imported_module = importlib.import_module(path)
-        imported_module.patch()
-
-    return on_import
-
+def integration_modname(intname, basemodname=None):
+    basemodname = basemodname or _BASE_MODULENAME
+    return '{}.{}'.format(basemodname, intname)
 
-def patch_all(**patch_modules):
-    """Automatically patches all available modules.
 
-    :param dict \**patch_modules: Override whether particular modules are patched or not.
+def install_all(overrides=None, raise_errors=False):
+    """
+    Installs all default enabled integrations allowing the defaults to be
+    overridden.
 
-        >>> patch_all(redis=False, cassandra=False)
+    :param overrides: particular integrations to override
+    :type overrides: dict
+    :param raise_errors: whether or not to raise errors when they occur
+    :type raise_errors: bool
+    """
+    overrides = overrides or {}
+    integrations_to_install = DEFAULT_INTEGRATIONS.copy()
+    integrations_to_install.update(overrides)
+
+    for integration, enabled in integrations_to_install.items():
+        if not enabled:
+            log.info(
+                'install: skipping disabled integration "{}"'.format(integration)
+            )
+            continue
+        install(integration, raise_errors)
+
+
+def install(integration, raise_errors=False):
+    """Installs an integration using its ``patch()`` function.
+
+    :param integration: the integration to install
+    :type integration: str
+    :param raise_errors: whether or not to raise errors if they occur
+    :type raise_errors: bool
     """
-    modules = PATCH_MODULES.copy()
-    modules.update(patch_modules)
+    intmodname = integration_modname(integration)
 
-    patch(raise_errors=False, **modules)
+    try:
+        intmod = importlib.import_module(intmodname)
+    except ImportError:
+        log_msg = 'install: integration "{}" not found'.format(integration)
+        if raise_errors:
+            raise PatchException(log_msg)
+        else:
+            log.error(log_msg)
+            return
 
+    if not hasattr(intmod, 'patch'):
+        log.error('install: integration "{}" does not have patch attribute'.format(integration))
+        return
 
-def patch(raise_errors=True, **patch_modules):
-    """Patch only a set of given modules.
+    intmod.patch()
```

Comment:
We should have a `try/except` around this.

Reply:
Great idea.